### PR TITLE
Add basic support for subordinate user/group ids

### PR DESCRIPTION
**Pull request by tiran**
New LDAP object class "ipaUserSubordinate" with four new fields:
- ipasubuidnumber / ipasubuidcount
- ipasubgidnumber / ipasgbuidcount

New self-service permission to add subids.

New command user-auto-subid to auto-assign subid

The code hard-codes counts to 65536, sets subgid equal to subuid, and
does not allow removal of subids. There is also a hack that emulates a
DNA plugin with step interval 65536 for testing.

## TODO

- [ ] add tooltip / doc to subordinate id UI. The term *subordinate id* is not a common term and may be confusing to admins or users.
- [ ] bike shed on command names ``user-auto-subid`` and ``user-match-subid``. Are these names self-explaining and easy to remember?
- [ ] Decide whether we need a command to remove subordinate ids from a user. I decided against it. Subid space is limited and we cannot safely recycle a subid any way.
- [ ] Include a helper script to import subuid and subgids from an existing databases? Customers may have existing ``/etc/subuid`` and ``/etc/subgid`` files.
- [ ] Enforce / check for subid interval overlaps? https://github.com/freeipa/freeipa/pull/5438 has some hacky code.
- [ ] Add helper to show how many subid intervals are left.
- [ ] Should we add a unique check on ``ipaSubUidNumber`` and ``ipaSubGidNumber``?
- [x] explain that IPA's random baseid does not conflict with subid range start.
- [x] Add example how subids are allocated and assign.
- [x] Investigate new ID range type. First attempt to add ``ipa-local-subid`` failed.
- [x] sidgen task fails ``Failed to convert LDAP entry to range struct`` in ``ipa_sidgen_common.c`` because subid_range has no RIDs. Added hack to include RIDs by default.
- [x] Failing sudo test case ``test_domain_resolution_order`` is tracked at https://github.com/SSSD/sssd/issues/5571
- [ ] submit OIDs to RHANANA
- [ ] Mention that systemd's dynamic user feature uses range uids in range 61184 to 65519.

See: https://pagure.io/freeipa/issue/8361
Signed-off-by: Christian Heimes <cheimes@redhat.com>

### CODE REVIEW

File: install/share/60basev3.ldif [hunk 1]
```diff
@@ -58,6 +58,10 @@ attributeTypes: (2.16.840.1.113730.3.8.11.70 NAME 'ipaPermTargetTo' DESC 'Destin
...
 attributeTypes: (2.16.840.1.113730.3.8.11.77 NAME 'ipaDomainResolutionOrder' DESC 'List of domains used to resolve a short name' EQUALITY caseIgnoreIA5Match SYNTAX 1.3.6.1.4.1.1466.115.121.1.26 SINGLE-VALUE X-ORIGIN 'IPA v4.5')
+attributeTypes: ( 2.16.840.1.113730.3.8.11.78 NAME 'ipaSubUidNumber' DESC 'numerical subordinate user ID' EQUALITY integerMatch ORDERING integerOrderingMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE  X-ORIGIN 'IPA v4.9')
+attributeTypes: ( 2.16.840.1.113730.3.8.11.79 NAME 'ipaSubUidCount' DESC 'numerical subordinate user ID count' EQUALITY integerMatch ORDERING integerOrderingMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE  X-ORIGIN 'IPA v4.9')
+attributeTypes: ( 2.16.840.1.113730.3.8.11.80 NAME 'ipaSubGidNumber' DESC 'numerical subordinate user ID' EQUALITY integerMatch ORDERING integerOrderingMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE  X-ORIGIN 'IPA v4.9')
+attributeTypes: ( 2.16.840.1.113730.3.8.11.81 NAME 'ipaSubGidCount' DESC 'numerical subordinate user ID count' EQUALITY integerMatch ORDERING integerOrderingMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE  X-ORIGIN 'IPA v4.9')
```
Comment (rcritten) [inline install/share/60basev3.ldif:64]: The OID base for v4 attributes is 2.16.840.1.113730.3.8.23.x

File: install/share/60basev3.ldif [hunk 2]
```diff
@@ -86,5 +90,6 @@ objectClasses: (2.16.840.1.113730.3.8.12.25 NAME 'ipaPrivateKeyObject' DESC 'Wra
 objectClasses: (2.16.840.1.113730.3.8.12.26 NAME 'ipaSecretKeyObject' DESC 'Wrapped secret keys' SUP top AUXILIARY MUST ( ipaSecretKey $ ipaWrappingKey $ ipaWrappingMech ) X-ORIGIN 'IPA v4.1' )
 objectClasses: (2.16.840.1.113730.3.8.12.34 NAME 'ipaSecretKeyRefObject' DESC 'Indirect storage for encoded key material' SUP top AUXILIARY MUST ( ipaSecretKeyRef ) X-ORIGIN 'IPA v4.1' )
 objectClasses: (2.16.840.1.113730.3.8.12.39 NAME 'ipaNameResolutionData' DESC 'Data used to resolve short names to fully-qualified form' SUP top AUXILIARY MAY ( ipaDomainResolutionOrder ) X-ORIGIN 'IPA v4.5')
+objectClasses: (2.16.840.1.113730.3.8.12.40 NAME 'ipaSubordinateId' DESC 'Subordinate uid and gid for users' SUP posixAccount AUXILIARY MUST ( ipaSubUidNumber $ ipaSubUidCount $ ipaSubGidNumber $ ipaSubGidCount ) X-ORIGIN 'IPA v4.9')
```
Comment (rcritten) [inline install/share/60basev3.ldif:93]: The OID base for v4 objectclasses is 2.16.840.1.113730.3.8.24.x

File: ACI.txt [hunk 3]
```diff
@@ -375,7 +375,7 @@ aci: (targetattr = "audio || businesscategory || carlicense || departmentnumber
 dn: dc=ipa,dc=example
```
Comment (rcritten) [inline ACI.txt:375]: You also need to update VERSION.m4 since you're adding new commands.

File: ipaserver/plugins/baseuser.py [hunk 4]
```diff
@@ -526,6 +567,82 @@ def convert_attribute_members(self, entry_attrs, *keys, **options):
...
+        else:
+            _entry_attrs = ldap.get_entry(dn, ["objectclass"])
+            entry_attrs["objectclass"] = _entry_attrs["objectclass"]
+            obj_classes = entry_attrs["objectclass"]
+
+        if "ipasubordinateid" not in obj_classes:
+            obj_classes.append("ipasubordinateid")
+
+        if subuid == DNA_MAGIC:
+            subuid = self._fake_dna_plugin(ldap, dn, entry_attrs)
```
Comment (rcritten) [inline ipaserver/plugins/baseuser.py:614]: I assume you'll want a comment around this too like XXX HACK

File: ipaserver/plugins/baseuser.py [hunk 5]
```diff
@@ -526,6 +567,82 @@ def convert_attribute_members(self, entry_attrs, *keys, **options):
...
+    def handle_subordinate_ids(self, ldap, dn, entry_attrs):
+        """Handle ipaSubordinateId object class
+        """
+        obj_classes = entry_attrs.get("objectclass")
+        new_subuid = entry_attrs.single_value.get("ipasubuidnumber")
+        new_subgid = entry_attrs.single_value.get("ipasubgidnumber")
+
+        # entry has object class ipaSubordinateId
+        # default to auto-assigment of subuids
+        if obj_classes is not None and "ipasubordinateid" in obj_classes:
```
Comment (rcritten) [inline ipaserver/plugins/baseuser.py:579]: There is a helper for this, self.obj.has_objectclass()

File: ipaserver/plugins/baseuser.py [hunk 6]
```diff
@@ -429,6 +435,41 @@ class baseuser(LDAPObject):
                     'J:', 'K:', 'L:', 'M:', 'N:', 'O:', 'P:', 'Q:', 'R:',
                     'S:', 'T:', 'U:', 'V:', 'W:', 'X:', 'Y:', 'Z:'),
                 ),
+        Int(
+            'ipasubuidnumber?',
+            label=_('SubUID range start'),
+            cli_name='subuid',
+            doc=_('Start value for subordinate user ID (subuid) range'),
+            minvalue=2 ** 31,
+            maxvalue=2 ** 32,
```
Comment (rcritten) [inline ipaserver/plugins/baseuser.py:444]: Use constants instead? SUBID_RANGE_START is one already defined.

**Comment by rcritten** [comment]
I haven't tested the patch yet but the direction looks ok.

I have a minor concern about adding four more lines to user* output. Do we need to show all the subuid attributes all the time? The user output already spans a lot as it is.

File: install/share/60basev3.ldif (hunk 1 above)
Comment (tiran) [reply install/share/60basev3.ldif]: Thanks! I changed all OIDs and moved the new ones to a new file ``60basev4.ldif``.

File: ACI.txt (hunk 3 above)
Comment (tiran) [reply ACI.txt:375]: I always forget about ``VERSION.m4``. Fixed!

File: ipaserver/plugins/baseuser.py (hunk 4 above)
Comment (tiran) [reply ipaserver/plugins/baseuser.py:615]: +1

File: ipaserver/plugins/baseuser.py (hunk 5 above)
Comment (tiran) [reply ipaserver/plugins/baseuser.py]: Thanks, that's an excellent suggestion.

File: ipaserver/plugins/baseuser.py (hunk 6 above)
Comment (tiran) [reply ipaserver/plugins/baseuser.py]: Good point!

**Comment by rcritten** [comment]
Patch seems to be working pretty well so far. I like the compromise to just show the SubUID range start by default on the cli.

Is there supposed to be uniqueness, and no overlap, between the subuids? It is not currently enforced. The subuid man page doesn't really say what it should do.

File: ipalib/constants.py [hunk 7]
```diff
@@ -340,3 +340,16 @@
...
+# subuid / subgid counts are hard-coded
+# An interval of 65536 uids/gids is required to map nobody (65534).
+SUBID_COUNT = 65536
+
+# upper half of uid_t (uint32_t)
+SUBID_RANGE_START = 2 ** 31
+# theoretical max limit is UINT32_MAX-1 ((2 ** 32) - 2)
+# We use a smaller value to keep the topmost subid interval unused.
+SUBID_RANGE_MAX = (2 ** 32) - (2 * SUBID_COUNT)
+SUBID_RANGE_SIZE = SUBID_RANGE_MAX - SUBID_RANGE_START
```
Comment (rcritten) [inline ipalib/constants.py:353]: This goes back to my previous lack of understanding of subuids. I'm asking this based on the assumption that userX gets range SUBUID_RANGE_START + SUBID_COUNT, userY gets range SUBID_RANGE_START + SUBID_COUNT + 1, etc.

So when we display range size in the idrange plugin how does SUBID_COUNT come into play?

**Comment by rcritten** [comment]
If possible it would be nice if SUBID_COUNT were used more globally. 65536 is still harcoded in places.

File: doc/designs/subordinate-ids.md [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+Some design choices are owed to the circumstance that uids and gids
+are limited datatypes. The Linux Kernel and userland defines
+``uid_t`` and ``gid_t`` as unsigned 32bit integers (``uint32_t``), which
+limits possible values for numeric user and group ids to
+``0 .. 2^32-2``. ``(uid_t)-1`` is reserved for error reporting. On the
+other hand the user ``nobody`` typically has uid 65534 / gid 65534. This
+means we need to assign 65,536 subordinate ids to every user. The
+theoretical maximum amount of subordinate ranges is less than 65,536
+(``65536 * 65536 == 2^32``). [``logins.def``](https://man7.org/linux/man-pages/man5/login.defs.5.html)
+also uses 65536 as default setting for ``SUB_UID_COUNT``.
```
Comment (rcritten) [inline doc/designs/subordinate-ids.md:45]: In code this is SUBID_COUNT.

File: ipalib/constants.py (hunk 7 above)
Comment (tiran) [reply ipalib/constants.py:353]: The ``count`` is the size of a subid range including the start value. The algorithm for automatic assignment would look like this:

```
subid_range_start = SUBID_RANGE_START + (n *  SUBID_COUNT)
n += 1
```

where ``n` is a counter starting at ``0``.


The first user owns the range ``[SUBID_RANGE_START, SUBID_RANGE_START + SUBUID_COUNT - 1]``. The second user owns the next batch of 65536 ids.

**Comment by rcritten** [comment]
Of the "selfservice: Add subordinate id" aci I'm trying to grok the del parts of it:

del=ipasubuidnumber:(!(ipasubuidnumber=*)) && ipasubuidcount:(!(ipasubuidcount=*)) && ipasubgidnumber:(!(ipasubgidnumber=*)) && ipasubgidcount:(!(ipasubgidcount=*))

Are you trying to prevent removal of the attributes? I think the objectclass will enforce their existence unless they are all removed at once (along with the objectclass).

**Comment by tiran** [comment]
> Of the "selfservice: Add subordinate id" aci I'm trying to grok the del parts of it:
> 
> `del=ipasubuidnumber:(!(ipasubuidnumber=*)) && ipasubuidcount:(!(ipasubuidcount=*)) && ipasubgidnumber:(!(ipasubgidnumber=*)) && ipasubgidcount:(!(ipasubgidcount=*))`
> 
> Are you trying to prevent removal of the attributes? I think the objectclass will enforce their existence unless they are all removed at once (along with the objectclass).

That's the infamous delete-when-empty check I mentioned in the comment above. https://github.com/389ds/389-ds-base/issues/4597 explains the gory details of the issue. Since IPA's LDAP driver uses ``MOD_REPLACE`` we have to allow deletes of non-existing entries.

**Comment by tiran** [comment]
Tests are now failing with:

```
2021-03-26T15:03:03Z DEBUG Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/ipaserver/install/adtrustinstance.py", line 411, in __add_rid_bases
    api.Backend.ldap2.modify_s(local_range.dn,
  File "/usr/lib/python3.9/site-packages/ipapython/ipaldap.py", line 855, in modify_s
    return self.conn.modify_s(dn, modlist)
  File "/usr/lib64/python3.9/site-packages/ldap/ldapobject.py", line 640, in modify_s
    return self.modify_ext_s(dn,modlist,None,None)
  File "/usr/lib64/python3.9/site-packages/ldap/ldapobject.py", line 613, in modify_ext_s
    resp_type, resp_data, resp_msgid, resp_ctrls = self.result3(msgid,all=1,timeout=self.timeout)
  File "/usr/lib64/python3.9/site-packages/ldap/ldapobject.py", line 764, in result3
    resp_type, resp_data, resp_msgid, decoded_resp_ctrls, retoid, retval = self.result4(
  File "/usr/lib64/python3.9/site-packages/ldap/ldapobject.py", line 774, in result4
    ldap_result = self._ldap_call(self._l.result4,msgid,all,timeout,add_ctrls,add_intermediates,add_extop)
  File "/usr/lib64/python3.9/site-packages/ldap/ldapobject.py", line 340, in _ldap_call
    reraise(exc_type, exc_value, exc_traceback)
  File "/usr/lib64/python3.9/site-packages/ldap/compat.py", line 46, in reraise
    raise exc_value
  File "/usr/lib64/python3.9/site-packages/ldap/ldapobject.py", line 324, in _ldap_call
    result = func(*args,**kwargs)
ldap.CONSTRAINT_VIOLATION: {'msgtype': 103, 'msgid': 39, 'result': 19, 'desc': 'Constraint violation', 'ctrls': [], 'info': 'New ID range has invalid type. All ranges in the same domain must be of the same type.'}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/ipaserver/install/service.py", line 635, in start_creation
    run_step(full_msg, method)
  File "/usr/lib/python3.9/site-packages/ipaserver/install/service.py", line 621, in run_step
    method()
  File "/usr/lib/python3.9/site-packages/ipaserver/install/adtrustinstance.py", line 418, in __add_rid_bases
    "object:\n  %s" % e[0]['info'])
TypeError: 'CONSTRAINT_VIOLATION' object is not subscriptable

2021-03-26T15:03:03Z DEBUG   [error] TypeError: 'CONSTRAINT_VIOLATION' object is not subscriptable
```

I guess the new range type was not a good idea in the first place.

File: ipalib/constants.py (hunk 7 above)
Comment (tiran) [reply ipalib/constants.py:353]: I have added additional documentation about subid allocation.

**Comment by tiran** [comment]
Multiple tests have failed due to DNS timeouts early in ipa-server-install setup phase.

**Comment by fcami** [comment]
While I would strongly object to [any compatibility breakage with older clients](https://github.com/SSSD/sssd/issues/5571), using ipa-ad-trust idrange type for subid range seems to allay my concerns.

**Comment by tiran** [comment]
The new variant with separate entries is implemented in PR https://github.com/freeipa/freeipa/pull/5713 and now ready for review.

**Pull request closed without merging**